spring-xd-lang-detector/src/main/java/org/springframework/xd/analytics/linguistics/langdetect/LanguageDetector.java

```diff
@@ -0,0 +1,336 @@
+
+		detector.append(text);
+
+		List<String> names = new ArrayList<String>();
+		List<Object> values = new ArrayList<Object>();
+
+		names.addAll(input.getFieldNames());
+		values.addAll(input.getValues());
+
+		if (isReturnMostLikelyLanguage()) {
+			names.add(getMostLikelyLanguageOutputPropertyName());
+			values.add(detector.detect());
+		}
+
+		if (isReturnLanguageProbabilities()) {
+			names.add(getLanguageProbabilitiesOutputPropertyName());
+			values.add(detector.getProbabilities());
+		}
+
+		return TupleBuilder.tuple().ofNamesAndValues(names, values);
+	}
+
+	/**
+	 * This creates a new {@link com.cybozu.labs.langdetect.Detector} instance.
+	 * We have to create the instance ourselves to avoid problems with the shared state inside {@link com.cybozu.labs.langdetect.DetectorFactory}.
+	 *
+	 * @param detectorFactoryState
+	 * @return
+	 */
+	private Detector newDetector(DetectorFactoryState detectorFactoryState) throws LangDetectException {
+
+		Detector detector = new Detector(detectorFactoryState.getWordLangProbMap(), detectorFactoryState.getLanguageList(), detectorFactoryState.getSeed());
+
+		if (!CollectionUtils.isEmpty(languagePriorityMap)) {
+			detector.setPriorMap(languagePriorityMap);
+		}
+
+		return detector;
+	}
+
+	private boolean isLanguageDetectionEnabled() {
+		return isReturnMostLikelyLanguage() || isReturnLanguageProbabilities();
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+
+		this.languagePriorityMap = new HashMap<String,Double>(languagePriorityParser.parseToLanguagePriorityMap(languagePriorities));
+
+		loadLanguageProfiles();
+
+		this.detectorFactoryState = captureDetectorFactoryState();
+
+		LOG.info("Loaded language profiles from {}.", languageProfileLocation);
+	}
+
+	private void loadLanguageProfiles() throws LangDetectException, IOException {
+
+		if (isEmpty(this.languageProfileLocation)) {
+			LOG.info("Using embedded language profiles.");
+			loadEmbeddedLangaugeProfiles();
+			return;
+		}
+
+		LOG.info("Using language profiles from {}.", languageProfileLocation);
+		loadExternalLanguageProfiles();
+	}
+
+	private void loadExternalLanguageProfiles() throws LangDetectException, IOException {
+		Resource languageProfileResource = new DefaultResourceLoader(getClass().getClassLoader()).getResource(languageProfileLocation);
+		DetectorFactory.loadProfile(languageProfileResource.getFile());
+	}
+
+	private void loadEmbeddedLangaugeProfiles() throws LangDetectException {
+		DetectorFactory.loadProfile(extractEmbeddedLanguageModels());
+	}
+
+	/**
+	 * Captures the current state of the {@link com.cybozu.labs.langdetect.DetectorFactory} to make sure
+	 * that state cannot be overridden by concurrent initializations later on.
+	 * <p>This is necessary since, the state in the {@code DetectorFactory} is stored globally.</p>
+	 *
+	 * @return
+	 */
+	@SuppressWarnings("unchecked")
+	private DetectorFactoryState captureDetectorFactoryState() {
```

Comment:
This may prove to be unnecessary, as those classes are loaded by the module class loader (hence 2 instances of the modules would get 2 instances of DetectorFactory.class)
I would not argue for its removal though.

Ironically, this is causing class leakage because of BeanInfo cache leakage, which ideally should be taken care of at the XD level